modules/linear_boltzmann_solvers/lbs_solver/io/flux_moments_io.cc

```diff
@@ -15,99 +16,71 @@ LBSSolverIO::WriteFluxMoments(
-                            "  uint64_t    group\n"
-                            "  double      value\n";
-
-  int header_size = (int)header_info.length();
-  char header_bytes[num_bytes];
-  memset(header_bytes, '-', num_bytes);
-  strncpy(header_bytes, header_info.c_str(), std::min(header_size, num_bytes - 1));
-  header_bytes[num_bytes - 1] = '\0';
-
-  file << header_bytes;
-
-  // Write macro data
+  const auto& grid = lbs_solver.Grid();
+  const auto& discretization = lbs_solver.SpatialDiscretization();
+
   const auto& uk_man = lbs_solver.UnknownManager();
   const auto NODES_ONLY = UnknownManager::GetUnitaryUnknownManager();
-  auto& discretization = lbs_solver.SpatialDiscretization();
-  auto& grid = lbs_solver.Grid();
-  const uint64_t num_local_cells = grid.local_cells.size();
-  const uint64_t num_local_nodes = discretization.GetNumLocalDOFs(NODES_ONLY);
-  const uint64_t num_moments = lbs_solver.NumMoments();
-  const uint64_t num_groups = lbs_solver.NumGroups();
-  const auto num_local_dofs = discretization.GetNumLocalDOFs(uk_man);
-  OpenSnLogicalErrorIf(src.size() != num_local_dofs, "Incompatible flux moments vector provided..");
 
-  file.write((char*)&num_local_cells, sizeof(uint64_t));
-  file.write((char*)&num_local_nodes, sizeof(uint64_t));
-  file.write((char*)&num_moments, sizeof(uint64_t));
-  file.write((char*)&num_groups, sizeof(uint64_t));
+  uint64_t num_local_cells = grid.local_cells.size();
+  uint64_t num_local_nodes = discretization.GetNumLocalDOFs(NODES_ONLY);
+  uint64_t num_moments = lbs_solver.NumMoments();
+  uint64_t num_groups = lbs_solver.NumGroups();
+  uint64_t num_local_dofs = num_local_nodes * num_moments * num_groups;
+  OpenSnLogicalErrorIf(src.size() != num_local_dofs, "Incompatible flux moments vector provided.");
 
-  // Write nodal positions
+  H5CreateAttribute(file_id, "num_moments", num_moments);
+  H5CreateAttribute(file_id, "num_groups", num_groups);
+
+  // Loop through mesh nodes and store data
+  std::vector<uint64_t> cell_ids, num_cell_nodes;
+  std::vector<double> nodes_x, nodes_y, nodes_z;
   for (const auto& cell : grid.local_cells)
   {
-    const uint64_t cell_global_id = cell.global_id;
-    const uint64_t num_cell_nodes = discretization.GetCellNumNodes(cell);
-
-    file.write((char*)&cell_global_id, sizeof(uint64_t));
-    file.write((char*)&num_cell_nodes, sizeof(uint64_t));
+    cell_ids.push_back(cell.global_id);
+    num_cell_nodes.push_back(discretization.GetCellNumNodes(cell));
 
     const auto nodes = discretization.GetCellNodeLocations(cell);
     for (const auto& node : nodes)
     {
-      file.write((char*)&node.x, sizeof(double));
-      file.write((char*)&node.y, sizeof(double));
-      file.write((char*)&node.z, sizeof(double));
-    } // for node
-  }   // for cell
-
-  // Write flux moments data
+      nodes_x.push_back(node.x);
+      nodes_y.push_back(node.y);
+      nodes_z.push_back(node.z);
+    }
+  }
+
+  // Write mesh data to h5
```

Comment:
maybe make the common more explciit that mesh is placed in its own group with group attributes

Reply:
Changed the comment